cpio/cpio.c

```diff
@@ -865,24 +904,8 @@ entry_to_archive(struct cpio *cpio, struct archive_entry *entry)
 	if (r == ARCHIVE_FATAL)
 		exit(1);
 
-	if (r >= ARCHIVE_WARN && archive_entry_size(entry) > 0 && fd >= 0) {
-		bytes_read = read(fd, cpio->buff, (unsigned)cpio->buff_size);
-		while (bytes_read > 0) {
-			ssize_t bytes_write;
-			bytes_write = archive_write_data(cpio->archive,
-			    cpio->buff, bytes_read);
-			if (bytes_write < 0)
-				lafe_errc(1, archive_errno(cpio->archive),
-				    "%s", archive_error_string(cpio->archive));
-			if (bytes_write < bytes_read) {
-				lafe_warnc(0,
-				    "Truncated write; file may have "
-				    "grown while being archived.");
-			}
-			bytes_read = read(fd, cpio->buff,
-			    (unsigned)cpio->buff_size);
-		}
-	}
+	if (r >= ARCHIVE_WARN && archive_entry_size(entry) > 0 && fd >= 0)
+		copy_file(cpio, fd);
```

Comment:
If we can line up the API right, this should just be a call to `archive_write_disk_copy_file`.  We should always prefer to have a clean public API and use that as directly as possible.

Looking at this more closely makes me wonder about the right structure here:
* The easiest API to explain would be one where you first call `archive_write_header` to write the header and then `archive_write_contents_from_file(archive, srcpath)` to write the body of that entry.
* But as the comments above explain, that complicates error handling; if the source file cannot be opened, we would like to be able to just skip this entry entirely (not writing the header).  That's why the original code was written to copy contents from a file descriptor.
* Exposing `copy_file()` from a file descriptor raises a lot of questions about the full meaning of that call (for example, what if the file descriptor is not at the beginning of the file?).

So I'm a little torn:  copying from a file descriptor enables the right kind of error handling, but copying from a filename is a cleaner API.  Thoughts?

Reply:
> Looking at this more closely makes me wonder about the right structure here:
> 
>    * The easiest API to explain would be one where you first call `archive_write_header` to write the header and then `archive_write_contents_from_file(archive, srcpath)` to write the body of that entry.

Agreed!

>    * But as the comments above explain, that complicates error handling; if the source file cannot be opened, we would like to be able to just skip this entry entirely (not writing the header).  That's why the original code was written to copy contents from a file descriptor.

Yeah, I re-wrote my code a bunch of times to try to handle every error path.  The API isn't perfect, I agree.  I guess it could use the srcpath instead of a file descriptor but, apart from the extra pain in error handling opening the file, the original file descriptor would be also required.  I.e. we would have to open that file twice: once for the header and the second time in the public API.

>    * Exposing `copy_file()` from a file descriptor raises a lot of questions about the full meaning of that call (for example, what if the file descriptor is not at the beginning of the file?).

Right, my implementation doesn't really care about that as it simply assumes the FDs (both of them) are in the right position.  The API will attempt to copy X bytes from the current fd_in to the current fd_out. 

> So I'm a little torn: copying from a file descriptor enables the right kind of error handling, but copying from a filename is a cleaner API. Thoughts?

I would rather not re-open the file in the API as this would mean yet another syscall.  But I understand your point.